packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,240 @@
+import { UrlWithParsedQuery } from 'url'
+import { IncomingMessage, ServerResponse } from 'http'
+import { join } from 'path'
+import { mediaType } from '@hapi/accept'
+import { createReadStream, promises } from 'fs'
+import { createHash } from 'crypto'
+import Server from './next-server'
+import { getContentType, getExtension } from './serve-static'
+import { fileExists } from '../../lib/file-exists'
+
+let sharp: typeof import('sharp')
+//const AVIF = 'image/avif'
+const WEBP = 'image/webp'
+const PNG = 'image/png'
+const JPEG = 'image/jpeg'
+const MIME_TYPES = [/* AVIF, */ WEBP, PNG, JPEG]
+const CACHE_VERSION = 1
+
+export async function imageOptimizer(
+  server: Server,
+  req: IncomingMessage,
+  res: ServerResponse,
+  parsedUrl: UrlWithParsedQuery
+) {
+  const { nextConfig, distDir } = server
+  const { sizes = [], domains = [] } = nextConfig?.experimental?.images || {}
+  const { headers } = req
+  const { url, w, q } = parsedUrl.query
+  const proto = headers['x-forwarded-proto'] || 'http'
+  const host = headers['x-forwarded-host'] || headers.host
+  const mimeType = mediaType(headers.accept, MIME_TYPES) || ''
+
+  if (!url) {
+    res.statusCode = 400
+    res.end('"url" parameter is required')
+    return { finished: true }
+  } else if (Array.isArray(url)) {
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+  } catch (_error) {
+    // url was not absolute so assuming relative url
+    try {
+      absoluteUrl = new URL(url, `${proto}://${host}`)
+    } catch (__error) {
+      res.statusCode = 400
+      res.end('"url" parameter is invalid')
+      return { finished: true }
+    }
+  }
+
+  if (!['http:', 'https:'].includes(absoluteUrl.protocol)) {
+    res.statusCode = 400
+    res.end('"url" parameter is invalid')
+    return { finished: true }
+  }
+
+  if (!server.renderOpts.dev && !domains.includes(absoluteUrl.hostname)) {
+    res.statusCode = 400
+    res.end('"url" parameter is not allowed')
+    return { finished: true }
+  }
```

Comment:
I think we need to validate that `absoluteUrl.pathname` is not `/_next/image/?` to avoid infinite loop.

Reply:
Actually, you need to pass parameters, so it cannot be infinite and not a big problem.